### ISSUE TITLE
Redundant closue lint can lose type annotation

### ISSUE THREAD

**Issue by neverRare** [rust-lang/rust-clippy#8548]
### Summary

Redundant closure lint can lose type annotation. This can lead to `type annotation needed` error.

### Lint Name

redundant_closure 

### Reproducer

In this rather contrived code

```rust
fn id<T>(x: T) -> T {
    x
}
fn main() {
    let _ = None.map(|x: i32| id(x));
}
```

Clippy reports the following

```
warning: redundant closure
 [--> src/main.rs:5:22
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021#)  |
5 |     let _ = None.map(|x: i32| id(x));
  |                      ^^^^^^^^^^^^^^ help: replace the closure with the function itself: `id`
  |
  = note: `#[warn(clippy::redundant_closure)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
```

But if I do replace it, I get the following error

```
error[[E0282]](https://doc.rust-lang.org/nightly/error-index.html#E0282): type annotations needed for `Option<T>`
 [--> src/lib.rs:5:13
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021#)  |
5 |     let _ = None.map(id);
  |         -   ^^^^ cannot infer type for type parameter `T` declared on the enum `Option`
  |         |
  |         consider giving this pattern the explicit type `Option<T>`, where the type parameter `T` is specified

For more information about this error, try `rustc --explain E0282`.
```

**Comment by profi248** [comment]
I've ran into this issue with `clippy --fix`, where Clippy replaced the closure with a direct mapping and broke the code.

This code:
```rust
pub async fn get_backup_path(&mut self) -> anyhow::Result<Option<PathBuf>> {
    let path = sqlx::query("select value from config where key = 'backup_path'")
        .fetch_optional(&mut self.transaction)
        .await?
        .map(|row| row.get(0));

    Ok(path.map(|path: String| PathBuf::from(path)))
}
```

Was "fixed" by Clippy by doing this:
```rust
pub async fn get_backup_path(&mut self) -> anyhow::Result<Option<PathBuf>> {
    let path = sqlx::query("select value from config where key = 'backup_path'")
        .fetch_optional(&mut self.transaction)
        .await?
        .map(|row| row.get(0));

    Ok(path.map(PathBuf::from))
}
```

Which produced this error:
```
error[E0282]: type annotations needed for `std::option::Option<_>`
  --> client/src/config/backup.rs:63:13
   |
63 |         let path = sqlx::query("select value from config where key = 'backup_path'")
   |             ^^^^
   |
help: consider giving `path` an explicit type, where the type for type parameter `U` is specified
   |
63 |         let path: std::option::Option<_> = sqlx::query("select value from config where key = 'backup_path'")
   |                 ++++++++++++++++++++++++

error[E0283]: type annotations needed for `std::option::Option<_>`
  --> client/src/config/backup.rs:63:13
   |
63 |         let path = sqlx::query("select value from config where key = 'backup_path'")
   |             ^^^^
...
68 |         Ok(path.map(PathBuf::from))
   |                     ------------- type must be known at this point
   |
   = note: multiple `impl`s satisfying `std::path::PathBuf: std::convert::From<_>` found in the following crates: `core`, `std`:
           - impl std::convert::From<std::boxed::Box<std::path::Path>> for std::path::PathBuf;
           - impl std::convert::From<std::ffi::OsString> for std::path::PathBuf;
           - impl std::convert::From<std::string::String> for std::path::PathBuf;
           - impl<'a> std::convert::From<std::borrow::Cow<'a, std::path::Path>> for std::path::PathBuf;
           - impl<T> std::convert::From<!> for T;
           - impl<T> std::convert::From<&T> for std::path::PathBuf
             where T: std::convert::AsRef<std::ffi::OsStr>, T: ?Sized;
           - impl<T> std::convert::From<T> for T;
help: consider giving `path` an explicit type, where the type for type parameter `T` is specified
   |
63 |         let path: std::option::Option<_> = sqlx::query("select value from config where key = 'backup_path'")
   |                 ++++++++++++++++++++++++

error[E0283]: type annotations needed for `std::option::Option<_>`
  --> client/src/config/backup.rs:63:13
   |
63 |         let path = sqlx::query("select value from config where key = 'backup_path'")
   |             ^^^^
...
68 |         Ok(path.map(PathBuf::from))
   |            ----------------------- type must be known at this point
   |
   = note: multiple `impl`s satisfying `std::path::PathBuf: std::convert::From<_>` found in the following crates: `core`, `std`:
           - impl std::convert::From<std::boxed::Box<std::path::Path>> for std::path::PathBuf;
           - impl std::convert::From<std::ffi::OsString> for std::path::PathBuf;
           - impl std::convert::From<std::string::String> for std::path::PathBuf;
           - impl<'a> std::convert::From<std::borrow::Cow<'a, std::path::Path>> for std::path::PathBuf;
           - impl<T> std::convert::From<!> for T;
           - impl<T> std::convert::From<&T> for std::path::PathBuf
             where T: std::convert::AsRef<std::ffi::OsStr>, T: ?Sized;
           - impl<T> std::convert::From<T> for T;
help: consider giving `path` an explicit type, where the type for type parameter `T` is specified
   |
63 |         let path: std::option::Option<_> = sqlx::query("select value from config where key = 'backup_path'")
   |                 ++++++++++++++++++++++++
```
